src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java

```diff
@@ -0,0 +1,199 @@
+        // Instantiation on the Jenkins Agents is done via deserialization.
+        JenkinsJVM.checkJenkinsJVM();
+    }
+
+    @Nonnull
+    @Override
+    public synchronized final PrintStream getLogger() {
+        if (logger == null) {
+            try {
+                logger = new PrintStream(new OtelLogOutputStream(buildInfo, w3cTraceContext, getLogEmitter(), clock), false, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new AssertionError(e);
+            }
+        }
+        return logger;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (logger != null) {
+            LOGGER.log(Level.FINE, () -> getClass().getName() + "#close(" + buildInfo + ")");
+            logger = null;
+        }
+        if (JenkinsJVM.isJenkinsJVM()) { // TODO why, it is possible because in the Agent closing of channel kill all the resources but on the Jenkins controller not.
+            OtelLogStorageFactory.get().close(buildInfo);
+        }
+    }
+
+    abstract LogEmitter getLogEmitter();
+
+    /**
+     * {@link OtelLogSenderBuildListener} implementation that runs on the Jenkins Controller and
+     * that retrieves the {@link LogEmitter} from the {@link OpenTelemetrySdkProvider}
+     */
+    static final class OtelLogSenderBuildListenerOnController extends OtelLogSenderBuildListener {
+        private static final long serialVersionUID = 1;
+
+        private final static Logger logger = Logger.getLogger(OtelLogSenderBuildListenerOnController.class.getName());
+
+        public OtelLogSenderBuildListenerOnController(@Nonnull BuildInfo buildInfo, @Nonnull Map<String, String> otelConfigProperties, @Nonnull Map<String, String> otelResourceAttributes) {
+            super(buildInfo, otelConfigProperties, otelResourceAttributes);
+            logger.log(Level.FINEST, () -> "new OtelLogSenderBuildListenerOnController()");
+            JenkinsJVM.checkJenkinsJVM();
+        }
+
+        @Override
+        public LogEmitter getLogEmitter() {
+            JenkinsJVM.checkJenkinsJVM();
+            return OpenTelemetrySdkProvider.get().getLogEmitter();
+        }
+
+        /**
+         * Java serialization to send the {@link OtelLogSenderBuildListener} from the Jenkins Controller to a Jenkins Agent.
+         * Swap the instance from a {@link OtelLogSenderBuildListenerOnController} to a {@link OtelLogSenderBuildListenerOnAgent}
+         * to change the implementation of {@link #getLogEmitter()}.
+         *
+         * See https://docs.oracle.com/en/java/javase/11/docs/specs/serialization/output.html#the-writereplace-method
+         */
+        private Object writeReplace() throws IOException {
+            logger.log(Level.FINEST, () -> "writeReplace()");
+            JenkinsJVM.checkJenkinsJVM();
+            return new OtelLogSenderBuildListenerOnAgent(buildInfo, otelConfigProperties, otelResourceAttributes);
```

Comment:
Is there any need to pass minimal credentials to the agent replacement as the CloudWatch Logs implementation does, or does the telemetry endpoint somewhat automatically verify the origin of a message so that a malicious build would not be able to spoof output from unrelated jobs?

Reply:
It answers my question by affirming that this implementation is indeed insecure for some configurations. https://github.com/jenkinsci/pipeline-cloudwatch-logs-plugin#security